src/sql/workbench/contrib/notebook/browser/notebookExplorer/notebookExplorerViewlet.ts

```diff
@@ -90,30 +125,273 @@ export class NotebookExplorerViewPaneContainer extends ViewPaneContainer {
+		}));
+
+		this._register(this.searchWidget.searchInput.onInput(() => this.searchView?.updateActions()));
+
+		this.trackInputBox(this.searchWidget.searchInputFocusTracker);
+	}
+
+	cancelSearch(focus: boolean = true): boolean {
+		if (focus) {
+			this.searchWidget.focus();
+			return true;
+		}
+		return false;
+	}
+
+	triggerQueryChange(_options?: { preserveFocus?: boolean, triggeredOnType?: boolean, delay?: number }) {
+		const options = { preserveFocus: true, triggeredOnType: false, delay: 0, ..._options };
+
+		if (!this.pauseSearching) {
+			this.triggerQueryDelayer.trigger(() => {
+				this._onQueryChanged(options.preserveFocus, options.triggeredOnType);
+			}, options.delay);
+		}
+
+		// initialize search view
+		if (!this.searchView) {
+			this.searchView = <NotebookSearchView>this.getView(NotebookSearchView.ID);
+		}
+	}
+
+	private _onQueryChanged(preserveFocus: boolean, triggeredOnType = false): void {
+		if (!this.searchWidget.searchInput.inputBox.isInputValid()) {
+			return;
+		}
+
+		const isRegex = this.searchWidget.searchInput.getRegex();
+		const isWholeWords = this.searchWidget.searchInput.getWholeWords();
+		const isCaseSensitive = this.searchWidget.searchInput.getCaseSensitive();
+		const contentPattern = this.searchWidget.searchInput.getValue();
+
+		if (contentPattern.length === 0) {
+			this.clearSearchResults(false);
+			this.updateViewletsState();
+			return;
+		}
+
+		const content: IPatternInfo = {
+			pattern: contentPattern,
+			isRegExp: isRegex,
+			isCaseSensitive: isCaseSensitive,
+			isWordMatch: isWholeWords
+		};
+
+		const excludePattern = undefined;
+		const includePattern = undefined;
+
+		// Need the full match line to correctly calculate replace text, if this is a search/replace with regex group references ($1, $2, ...).
+		// 10000 chars is enough to avoid sending huge amounts of text around, if you do a replace with a longer match, it may or may not resolve the group refs correctly.
+		// https://github.com/Microsoft/vscode/issues/58374
+		const charsPerLine = content.isRegExp ? 10000 : 1000;
+
+		const options: ITextQueryBuilderOptions = {
+			_reason: 'searchView',
+			extraFileResources: this.instantiationService.invokeFunction(getOutOfWorkspaceEditorResources),
+			maxResults: NotebookExplorerViewPaneContainer.MAX_TEXT_RESULTS,
+			disregardIgnoreFiles: undefined,
+			disregardExcludeSettings: undefined,
+			excludePattern,
+			includePattern,
+			previewOptions: {
+				matchLines: 1,
+				charsPerLine
+			},
+			isSmartCase: this.searchConfig.smartCase,
+			expandPatterns: true
+		};
+
+		const onQueryValidationError = (err: Error) => {
+			this.searchWidget.searchInput.showMessage({ content: err.message, type: MessageType.ERROR });
+			this.searchView.clearSearchResults();
+		};
+
+		let query: ITextQuery;
+		try {
+			query = this.queryBuilder.text(content, [], options);
+		} catch (err) {
+			onQueryValidationError(err);
+			return;
+		}
+
+		this.validateQuery(query).then(async () => {
```

Comment:
Why are you using async and then at the same time? This code is also having consequences after its execution because of this - it's probably better to use await here.